src/modules/commander/mag_calibration.cpp

```diff
@@ -626,6 +626,138 @@ calibrate_return mag_calibrate_all(orb_advert_t *mavlink_log_pub, int32_t cal_ma
+		param_get(param_find("CAL_MAG_ROT_AUTO"), &param_cal_mag_rot_auto);
+
+		if ((worker_data.calibration_sides >= 3) && (param_cal_mag_rot_auto == 1)) {
+
+			// find first internal mag to use as reference
+			int internal_index = -1;
+
+			for (unsigned cur_mag = 0; cur_mag < MAX_MAGS; cur_mag++) {
+				if (!worker_data.calibration[cur_mag].external() && (worker_data.calibration[cur_mag].device_id() != 0)) {
+					internal_index = cur_mag;
+					break;
+				}
+			}
+
+			// only proceed if there's a valid internal
+			if (internal_index >= 0) {
+
+				// apply new calibrations to all raw sensor data before comparison
+				for (unsigned cur_mag = 0; cur_mag < MAX_MAGS; cur_mag++) {
+					if (worker_data.calibration[cur_mag].device_id() != 0) {
+						for (unsigned i = 0; i < worker_data.calibration_counter_total[cur_mag]; i++) {
+
+							const Vector3f sample{worker_data.x[cur_mag][i], worker_data.y[cur_mag][i], worker_data.z[cur_mag][i]};
+
+							float scale_data[9] {
+								diag[cur_mag](0),    offdiag[cur_mag](0), offdiag[cur_mag](1),
+								offdiag[cur_mag](0),    diag[cur_mag](1), offdiag[cur_mag](2),
+								offdiag[cur_mag](1), offdiag[cur_mag](2),    diag[cur_mag](2)
+							};
+
+							// apply calibration
+							const Vector3f m{Matrix3f{scale_data} *(sample - sphere[cur_mag])};
+
+							// store back in worker_data
+							worker_data.x[cur_mag][i] = m(0);
+							worker_data.y[cur_mag][i] = m(1);
+							worker_data.z[cur_mag][i] = m(2);
+						}
+					}
+				}
+
+				// rotate internal mag data to board
+				const Dcmf board_rotation = calibration::GetBoardRotation();
+
+				for (unsigned i = 0; i < worker_data.calibration_counter_total[internal_index]; i++) {
+
+					const Vector3f m = board_rotation * Vector3f{worker_data.x[internal_index][i],
+							worker_data.y[internal_index][i], worker_data.z[internal_index][i]};
+
+					worker_data.x[internal_index][i] = m(0);
+					worker_data.y[internal_index][i] = m(1);
+					worker_data.z[internal_index][i] = m(2);
+				}
+
+				for (unsigned cur_mag = 0; cur_mag < MAX_MAGS; cur_mag++) {
+					if (worker_data.calibration[cur_mag].external() && (worker_data.calibration[cur_mag].device_id() != 0)) {
+
+						const int last_sample_index = math::min(worker_data.calibration_counter_total[internal_index],
+											worker_data.calibration_counter_total[cur_mag]);
+
+						float diff_sum[ROTATION_MAX] {};
+
+						float min_diff = FLT_MAX;
+						Rotation best_rotation = ROTATION_NONE;
+
+						for (int r = ROTATION_NONE; r < ROTATION_MAX; r++) {
+							for (int i = 0; i < last_sample_index; i++) {
+
+								float x = worker_data.x[cur_mag][i];
+								float y = worker_data.y[cur_mag][i];
+								float z = worker_data.z[cur_mag][i];
+								rotate_3f((enum Rotation)r, x, y, z);
+
+								Vector3f diff = Vector3f{x, y, z} - Vector3f{worker_data.x[internal_index][i], worker_data.y[internal_index][i], worker_data.z[internal_index][i]};
+
+								diff_sum[r] += diff.norm();
```

Comment:
Instead of computing the norm (containing a sqrt) for each sample, we could sum the diff^2 (i.e.:`diff_sum[r] += diff.norm_squared()`) in the loop, and then divide by `3*last_sample_index` and compute the sqrt at the end to get the RMS value.

Reply:
Good idea, I've updated it to simply compute and compare MSE.